### PR TITLE
Dynamic log prefixes for `app dev` output

### PR DESCRIPTION
**Pull request by nickwesselman**
### WHY are these changes introduced?

Enables extension-specific prefixes in the `app dev` output. This improves the existing `app dev` output for extensions by allowing log prefixes to match the extension handle, which is already the case for `app build`. It also enables extension-specific prefixes in the forthcoming app/function log streaming.

### WHAT is this pull request doing?

* `ConcurrentOutput` now looks for a prefix stashed in an `AsyncLocalStorage`, which it will use instead of the process prefix.
* `ConcurrentOutput` allows for prefix column width to be set externally and not just on known processes.
* Theme extension process has been given a new prefix since it was the only thing left outputting `extensions`.
* Implements some UX recommendations on (less) use of color in `app dev` output to better confirm to our UX principles for Shopify CLI.

The colors used in `app dev` output weren't changed, but we might consider adding to them in a follow up PR.

#### Before

CLI app dev before extension prefixes

#### After

CLI app dev after extension prefixes

### How to test your changes?

The best test is running `app dev` against an app with many extensions. I have one here:
https://github.com/Shopify/nw-scratch/tree/many-extension-app

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [x] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx [hunk 1]
```diff
@@ -77,43 +82,84 @@ const ConcurrentOutput: FunctionComponent<ConcurrentOutputProps> = ({
...
+  }
+
+  const writableStream = (process: OutputProcess, prefixes: string[]) => {
+    return new Writable({
+      write(chunk, _encoding, next) {
+        const parsedLog : ParsedLog = parseLog(chunk.toString('utf8'))
+        const prefix = parsedLog.prefix ?? process.prefix
+        const index = addPrefix(prefix, prefixes)
+
+        const lines = stripAnsi(parsedLog.log.replace(/(\n)$/, '')).split(/\n/)
```
Comment (isaacroldan) [inline packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:139]: So we are parsing before splitting lines, do we assume chunk === line?
Could we receive log lines from different functions in the same chunk?

Comment (nickwesselman) [reply packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:137]: 🤔 I don't think so? Though we could add support for that later I think by looking for the prefix marker at the start of any line?

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx [hunk 2]
```diff
@@ -71,49 +77,95 @@ function currentTime() {
...
-  const prefixColumnSize = Math.max(...processes.map((process) => process.prefix.length))
-  const concurrentColors: TextProps['color'][] = useMemo(() => ['yellow', 'cyan', 'magenta', 'green', 'blue'], [])
-  const lineColor = useCallback(
-    (index: number) => {
-      const colorIndex = index < concurrentColors.length ? index : index % concurrentColors.length
-      return concurrentColors[colorIndex]!
-    },
-    [concurrentColors],
-  )
-  const writableStream = useCallback(
```
Comment (nickwesselman) [inline packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:89]: @isaacroldan The use of `useCallback` here seemed unnecessary and I ended up removing it while I was iterating on state management for the seen prefixes. I could probably reintroduce it now, but is it actually needed?

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx [hunk 3]
```diff
@@ -71,49 +77,95 @@ function currentTime() {
  */
 const ConcurrentOutput: FunctionComponent<ConcurrentOutputProps> = ({
   processes,
+  prefixColumnSize = 12,
```
Comment (nickwesselman) [inline packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:80]: What should our default minimum be?

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx [hunk 4]
```diff
@@ -71,49 +77,95 @@ function currentTime() {
...
-            },
-          ])
-          next()
-        },
-      })
-    },
-    [lineColor],
-  )
+
+  const concurrentColors = [
```
Comment (nickwesselman) [inline packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:88]: 🥚 Easter egg idea: Override colors with an environment variable

**Comment by nickwesselman** [comment]
Needs unit tests

**Comment by nickwesselman** [comment]
Need to review other uses of ConcurrentOutput as well

**Comment by nickwesselman** [comment]
A couple linting fixes for non-related work ended up in here as a result of a `lint:fix` as well.

File: packages/app/src/cli/services/dev/processes/draftable-extension.ts [hunk 5]
```diff
@@ -37,24 +37,48 @@ export const pushUpdatesForDraftableExtensions: DevProcessFunction<DraftableExte
...
+      const extensionErr = extension.getPrefixedLogger(stderr)
+
+      await extension.build({
+        app,
+        stdout: extensionOut,
+        stderr: extensionErr,
+        useTasks: false,
+        signal,
+        environment: 'development',
+      })
```
[the remark this answers is not in the document]
Comment (nickwesselman) [reply packages/app/src/cli/services/dev/processes/draftable-extension.ts:50]: I tried to keep the insertion of the extension-specific logger at the same level as much as possible. In this case, if it was done inside build, `pushUpdatesForDraftableExtensions` would need to know to _not_ wrap stdio streams for `extension.build`, but would still need to do it for `updateExtensionDraft` and `setupExtensionWatcher`. This actually seems cleaner to me? But definitely subjective.

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx (hunk 2 above)
Comment (nickwesselman) [reply packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:89]: Linting encouraged me to add it back... I believe that using it improves rendering performance.

File: packages/app/src/cli/models/extensions/extension-instance.ts [hunk 6]
```diff
@@ -227,6 +229,15 @@ export class ExtensionInstance<TConfiguration extends BaseConfigType = BaseConfi
...
   }
 
+  getPrefixedLogger(wrappedWritable: Writable) {
+    const handle = this.handle
+    return new Writable({
+      write(chunk, _encoding, callback) {
+        wrappedWritable.write(`${prefixConcurrentOutputLog(handle, chunk.toString('utf8'))}`, 'utf8', callback)
+      },
+    })
+  }
```
Comment (isaacroldan) [inline packages/app/src/cli/models/extensions/extension-instance.ts:239]: So we are creating multiple Writable objects, but in reality we are writing to the same one and using this as a "proxy". Does this has any downsides?

File: packages/app/src/cli/services/dev/update-extension.ts [hunk 7]
```diff
@@ -64,7 +64,7 @@ export async function updateExtensionDraft({
     stderr.write(`Error while updating drafts: ${errors}`)
   } else {
     const draftUpdateSuccesMessage = extension.draftMessages.successMessage
-    if (draftUpdateSuccesMessage) outputInfo(draftUpdateSuccesMessage, stdout)
+    if (draftUpdateSuccesMessage) outputInfo(`${draftUpdateSuccesMessage}`, stdout)
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev/update-extension.ts:67]: do you need this change if `draftUpdateSuccesMessage` is already a string?

File: packages/app/src/cli/services/dev/extension/payload/store.ts [hunk 8]
```diff
@@ -37,7 +37,15 @@ export async function getExtensionsPayloadStoreRawPayload(
...
-    extensions: await Promise.all(options.extensions.map((extension) => getUIExtensionPayload(extension, options))),
+    extensions: await Promise.all(
+      options.extensions.map((extension) =>
+        getUIExtensionPayload(extension, {
+          ...options,
+          stdout: extension.getPrefixedLogger(options.stdout),
+          stderr: extension.getPrefixedLogger(options.stderr),
+        }),
+      ),
+    ),
```
Comment (isaacroldan) [inline packages/app/src/cli/services/dev/extension/payload/store.ts:48]: Why do we need to include this in the payload? The only place I could find that we need a stdout is in `getLocalization`, correct? seems weird to be injecting a custom stdout to they payload store just for that 🤔 

I think this is another case where for me it makes sense that the extension itself decides to use the custom Writable 🤔

File: packages/app/src/cli/services/dev/processes/draftable-extension.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickwesselman) [reply packages/app/src/cli/services/dev/processes/draftable-extension.ts]: Resolved with the switch to AsyncLocalStorage.

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx (hunk 1 above)
Comment (nickwesselman) [reply packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx]: No longer relevant with the switch to AsyncLocalStorage.

File: packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx [hunk 9]
```diff
@@ -36,6 +38,16 @@ function currentTime() {
   return `${hours}:${minutes}:${seconds}`
 }
 
+interface ConcurrentOutputContext {
+  outputPrefix: string
+}
```
[the remark this answers is not in the document]
Comment (nickwesselman) [reply packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx:43]: https://github.com/Shopify/develop-app-management/issues/1801

**Pull request merged**